### PR TITLE
feat(dashboard): add table component support

### PR DESCRIPTION
**Pull request by square-li**
## Overview
add `UpdateTableAssetsAction` action to create props that table widget needs.
image
 

## Legal
This project is available under the [Apache 2.0 License](http://www.apache.org/licenses/LICENSE-2.0.html).

### CODE REVIEW

File: packages/dashboard/src/store/actions/updateTableWidget/index.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+        },
+      };
+      if (!propertyNames.includes(name)) {
+        propertyNames.push(name);
+      }
+    });
+    return item;
+  });
+  const columnDefinitions: ColumnDefinition[] = propertyNames.map((name) => ({ key: name, header: name }));
+  const updateWidget: (widgets: Widget) => Widget = (w) => {
```
Comment (tjuranek) [inline packages/dashboard/src/store/actions/updateTableWidget/index.ts:48]: How is this passing? Also I don't think there's a need to be explicit on the type for this?

`const updateWidget = (widgets: Widget[]) => {}` should be fine here.

File: packages/dashboard/src/store/actions/updateTableWidget/index.ts [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+    if (w.id === widget.id) {
+      return {
+        ...w,
```
Comment (tjuranek) [inline packages/dashboard/src/store/actions/updateTableWidget/index.ts:51]: Nit comment, but I'd avoid single char variables names unless it's absolutely beneficial.

File: packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx [hunk 3]
```diff
@@ -0,0 +1,154 @@
+import { AppKitWidget, Widget } from '~/types';
+import { DashboardState, initialState } from '~/store/state';
+import { MOCK_KPI_WIDGET } from '../../../../testing/mocks';
```
Comment (tjuranek) [inline packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx:3]: There a `~` way to import this?

File: packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx [hunk 4]
```diff
@@ -0,0 +1,154 @@
...
+    updateTableAssets(
+      state,
+      onUpdateTableAssets({
+        widget: MOCK_KPI_WIDGET,
+        assetQuery: MOCK_ASSET_QUERY,
+      })
+    )
+  ).toEqual(state);
+});
+it('should create item and column definitions', () => {
```
Comment (tjuranek) [inline packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx:140]: Should there also be tests for deleting items or columns?

**Review by tjuranek — CHANGES_REQUESTED** [review]

File: packages/dashboard/package.json [hunk 5]
```diff
@@ -99,6 +99,7 @@
     "@iot-app-kit/core": "^2.6.4",
```
[the remark this answers is not in the document]
Comment (square-li) [reply packages/dashboard/package.json:99]: Didn't know that. Thanks! Adding it.

File: packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx (hunk 3 above)
Comment (square-li) [reply packages/dashboard/src/store/actions/updateTableWidget/index.spec.tsx:3]: Updating

File: packages/dashboard/src/store/actions/updateTableWidget/index.ts [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+export const onUpdateTableAssets = (payload: UpdateTableAssetPayload): UpdateTableAssetsAction => ({
+  type: 'UPDATE_TABLE_ASSET',
+  payload,
+});
+
+export const updateTableAssets = (state: DashboardState, action: UpdateTableAssetsAction): DashboardState => {
+  const { assetQuery, widget } = action.payload;
+  if (widget.componentTag !== 'iot-table') return state;
```
[the remark this answers is not in the document]
Comment (square-li) [reply packages/dashboard/src/store/actions/updateTableWidget/index.ts:25]: Currently it is hardcoded [here](https://github.com/awslabs/iot-app-kit/blob/63bb953b55e3aa86122fa0b3958fe2742a28629e/packages/dashboard/src/types.ts) as
```
export const AppKitComponentTags = [
  'iot-bar-chart',
  'iot-kpi',
  'iot-line-chart',
  'iot-scatter-chart',
  'iot-status-grid',
  'iot-status-timeline',
  'iot-table',
] as const;
export type AppKitComponentTag = typeof AppKitComponentTags[number];
```
Typescript will warn if you provides a bad tag name.

File: packages/dashboard/src/store/actions/updateTableWidget/index.ts [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+  type: 'UPDATE_TABLE_ASSET';
+  payload: UpdateTableAssetPayload;
+}
+
+export const onUpdateTableAssets = (payload: UpdateTableAssetPayload): UpdateTableAssetsAction => ({
+  type: 'UPDATE_TABLE_ASSET',
+  payload,
+});
+
+export const updateTableAssets = (state: DashboardState, action: UpdateTableAssetsAction): DashboardState => {
```
Comment (jmbuss) [inline packages/dashboard/src/store/actions/updateTableWidget/index.ts:23]: I know this is the pattern we have been doing for other types of widgets with asset queries, but is there any way we can move this logic into the table component? I'm in the process of removing any store actions that don't apply to every widget. I don't want to add onto that work unless we absolutely have to.

**Comment by square-li** [comment]
Closing this PR due to ongoing refectoring. Will open another one targeting to the latest structure.

**Pull request closed without merging**